rust/lance/src/index/vector/ivf/builder.rs

```diff
@@ -165,7 +241,49 @@ pub(super) async fn build_partitions(
         precomputed_partitons,
     )?;
 
-    let shuffled = shuffle_dataset(data, column, ivf_model, pq.num_sub_vectors()).await?;
+    let (row_id_buffers, pq_code_buffers) = shuffle_dataset_v2(
+        data,
+        column,
+        ivf_model,
+        ivf.num_partitions() as u32,
+        pq.num_sub_vectors(),
+    )
+    .await?;
+
+    let shuffled = stream::iter(
+        row_id_buffers
+            .into_iter()
+            .zip(pq_code_buffers.into_iter())
+            .enumerate(),
+    )
+    .map(|(idx, (row_ids, pq_codes))| {
+        let schema = Arc::new(Schema::new(vec![
+            ROW_ID_FIELD.clone(),
+            Field::new(
+                PQ_CODE_COLUMN,
+                DataType::FixedSizeList(
+                    Arc::new(Field::new("item", DataType::UInt8, true)),
+                    pq.num_sub_vectors() as i32,
+                ),
+                false,
+            ),
+        ]));
+
+        let batch = RecordBatch::try_new(
+            schema,
+            vec![
+                Arc::new(UInt64Array::from_iter_values(row_ids.into_iter())),
+                Arc::new(FixedSizeListArray::new(
```

Comment:
You can use help function https://docs.rs/lance-arrow/0.8.21/lance_arrow/trait.FixedSizeListArrayExt.html#tymethod.try_new_from_values